### PR TITLE
[2.0] Extract "operation" classes

### PR DESCRIPTION
**Pull request by franzliedke**
This is my biggest one yet, and the last major step of refactoring. This was fun. :smile: 

**Some notes first:**
- I haven't yet rebased on your latest changes, as that will cause some merge conflicts. I wanted to give you the chance to review first.
- I noticed at some place in the middle of this refactoring that I automatically switched to PSR-2 code style, sorry about that. I will fix that once we have the features sorted.
- There are currently three failing tests. This is **intended** and explained further below.

Now, for the description of the actual changes:

1. Most importantly, I separated the configuration from the actual application of new headers, safe mode etc., as proposed in the initial discussion that led to v2.0. Things now work like this:

    a. The user configures security headers by using the `SecureHeaders` facade class, and the known methods (e.g. `csp`, `safeMode` etc.).
    b. When calling `done` (or later, `apply` with an adapter), the configuration is turned into a sequence of *operations* - things like "add a header, if it wasn't sent yet", "add flags to all cookies that match certain names" etc.
    c. `apply` retrieves the currently sent headers (in a `HeaderBag`) from the adapter, applies all operations, and then instructs the adapter to replace all headers with the ones from the now-modified `HeaderBag`.

2. I propose removing the `addHeader` and `header` methods. They make things more complicated than necessary (because their replace/add behavior is different than for headers added implicity, e.g. by configuring CSP directives). Also, it makes the whole architecture simpler because headers can just be sent the standard way (header() method for the global adapter, modifying the response for the PSR-7 adapter), and things like "safe mode" etc. will still be applied to them, without needing to store them internally in the first place. That's why there are currently three failing tests - I simply propose to remove these tests.

3. Testing is really comfortable now. The operations can be tested one-by-one, which lets us specify very clearly the edge cases that all operations need to be able to handle (case sensitivity, spacing, behavior with existing headers etc.). I have started on some of them, but would like to ask you to either add some tests, or specify what other behavior needs to be tested - you simply know more about the intended feature set than I do.

---

All in all, this makes the architecture much simpler: Configure, apply the changes, and send the headers. This makes it more clear when to do things like parsing attributes. Existing headers are simply ignored - but still sent - (without having to be inspected), as long as they are not in any way relevant to the security features.

### CODE REVIEW

**Comment by franzliedke** [comment]
I am sure that some things are still broken, but let's specify them through unit tests first.

**Comment by aidantwoods** [comment]
Woah, this is a big one – going to take some time to review!

Liking the changes overall so far though! 😄

File: src/SecureHeaders.php [hunk 1]
```diff
@@ -881,22 +875,114 @@ public function hpkproSubdomains($mode = true)
 
     public function done()
     {
-        $this->importHeaders();
-        $this->applyAutomaticHeaders();
+        // TODO: Move this elsewhere!
+        $directive = $this->canInjectStrictDynamic();
+        if ( ! is_int($directive)) {
```
Comment (aidantwoods) [inline src/SecureHeaders.php]: Should add `and is_string($directive)` to the initial `if` here (`canInjectStrictDynamic` will return the bool false if it the checks allowing it to inject fail, (integer return is an ~~error~~ exit code), and string with the directive name if it can).

I think this was previously just `and $directive`, seeing now that probably wasn't clear enough ;p so `is_string` should do it?

File: src/SecureHeaders.php [hunk 2]
```diff
@@ -881,22 +875,114 @@ public function hpkproSubdomains($mode = true)
...
+            );
+        }
+
+        $this->apply($this->httpAdapter);
+
+        // TODO: Take care of these!
+        #$this->reportMissingHeaders();
+        #$this->validateHeaders();
+        #$this->reportErrors();
+    }
```
Comment (aidantwoods) [inline src/SecureHeaders.php]: Indeed 😉 

The header validation will hinge on finishing up the header parse.
Quite impressed with the attribute parsing/writing from what I've read so far though – you've managed to condense most of the functionality into something a lot more readable than the conditional regular expressions I was using to parse each header into attributes 😉 

I guess the only thing missing is decomposing the CSP (and CSPRO) header into directive name/source list string pairs (this to say that it'll need special casing for the parse). I'll try finish up reading through all the changes, and then I can probably help out with some commits if you haven't already added it in by then 😜

**Comment by aidantwoods** [comment]
> I am sure that some things are still broken, but let's specify them through unit tests first.

Agreed 👍

File: src/SecureHeaders.php [hunk 3]
```diff
@@ -881,22 +875,114 @@ public function hpkproSubdomains($mode = true)
...
+
+    /**
+     * Return an array of header operations, depending on current configuration.
+     *
+     * These can then be applied to e.g. the current set of headers.
+     *
+     * @return Operation[]
+     */
+    private function pipeline()
```
Comment (aidantwoods) [inline src/SecureHeaders.php:893]: Really like this 😄 👍

File: src/Operation.php [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+interface Operation
+{
+    /**
+     * Transform the given set of headers
+     *
+     * @param HeaderBag $headers
+     * @return void
+     */
+    public function modify(HeaderBag $headers);
```
Comment (aidantwoods) [inline src/Operation.php]: Just for clarity here, shall we include the an ampersand to make clear that the object passed can be modified 😉 

As in changing to `public function modify(HeaderBag &$headers);`

(The PHP [docs lie](https://php.net/manual/functions.arguments.php) and say that arguments are passed by value, so the fact that `$headers` object can be (and is) modified (in the outer scope) might easily be overlooked)

**Comment by aidantwoods** [comment]
Okay, so provided you're not already working on more changes? I can just add in the new features I added previously on top of your changes (we can then just prefer the changes in this PR to resolve the conflicts).

That's it from me tonight though, I'll take a look at adding in those new features tomorrow

**Comment by franzliedke** [comment]
I've fixed the header validation now.

Can you comment on these two questions:
1. Is it okay to remove `addHeader()` and `header()` methods?
2. Can we get rid of PHP 5.3 support for v2.0? If I am not mistaken, I have used some 5.4-only features like `$this` in callbacks.

File: src/Operation.php (hunk 4 above)
Comment (franzliedke) [reply src/Operation.php]: No, not necessary. Objects are de-facto passed by reference. Technically, it is still passing by value, see the first comment on [this SO answer](http://stackoverflow.com/a/881/192741) for the more correct explanation.

Comment (aidantwoods) [reply src/Operation.php]: I know it's not necessary to work, my point is it's more clear for the reader ;p

(The fact that the actual behaviour can only be determined by tests or reading [stack overflow](http://stackoverflow.com/a/9696799) answers makes PHPs object 'value' interpretation pretty uncommon knowledge)

Comment (franzliedke) [reply src/Operation.php]: Seeing the code of other PHP libraries, I'd say the knowledge is very common in PHP land.

Also, passing objects by value could get super expensive with deep cloning etc., so I don't think this is really a problem.

**Comment by aidantwoods** [comment]
> 1. Is it okay to remove addHeader() and header() methods?

Yup, that's fine :)

> 2. Can we get rid of PHP 5.3 support for v2.0? If I am not mistaken, I have used some 5.4-only features like $this in callbacks.

I think that'll be okay, if it's going to cause lots of code changes. Can always refer to the 1.x versions if people need this on PHP5.3

File: src/Operation.php (hunk 4 above)
Comment (aidantwoods) [reply src/Operation.php]: I'm not so sure, the two top answers (including the accepted answer) in that SO link you posted are both wrong ;p

Surely this is just a notational change? (since the use case here is identical to a reference)

Comment (franzliedke) [reply src/Operation.php]: Well, it's just super uncommon in PHP code I've seen. In the end, it's your decision. ;)

**Comment by franzliedke** [comment]
Okay, I've removed the `addHeader()` stuff and upped the PHP requirement to v5.4.
The methods like `importHSPS()` and `importHPKP()` are now unused - do you want to make them public as an alternative for importing directives? Else, they should be removed.

Also, behavior of `CompileCSP` and similar operations needs to be tested for cases when the header has already been sent another way.

File: src/Operation.php (hunk 4 above)
Comment (aidantwoods) [reply src/Operation.php]: I usually prefer to be as explicit as possible with these things ;-)

That and I feel that the PHP docs are very misleading with regard to how objects are passed. (Only true mention of this behaviour that I could find is contained in the equivalent of some change [notes](http://www.php.net/manual/en/migration5.oop.php), while the page on [function arguments](https://php.net/manual/functions.arguments.php) flat out contradicts their explaination.

**Comment by aidantwoods** [comment]
For `importHSTS` and `importHPKP`, I think these are probably not so useful (they're relatively simple directives) so we can probably just get rid of these.

I'd still like to keep the behaviour with `importCSP` though, I've added some changes to the `CompileCSP` operation that should support either merging all headers with appropriate names or sending the headers side by side (can make this configurable through `SecureHeaders` via a setter at some point).

This should address what `CompileCSP` should do when encountering other headers (it definitely shouldn't remove them IMO).

**Comment by franzliedke** [comment]
If I remember correctly, the previous behavior was that the configured directives were simply ignored when the header was already sent from somewhere else...

File: src/HeaderBag.php [hunk 5]
```diff
@@ -86,6 +86,24 @@ function ($all, $item) {
...
+            return array();
+        }
+
+        return array_reduce(
+            array($this->headers[$name]),
+            function ($all, $item) {
+                return array_merge($all, $item);
+            },
+            array()
+        );
```
Comment (franzliedke) [inline src/HeaderBag.php]: Shouldn't `return $this->headers[$name]` here be enough?

Comment (aidantwoods) [reply src/HeaderBag.php]: Oops, yeah... definitely aha

**Comment by aidantwoods** [comment]
> the previous behavior was that the configured directives were simply ignored when the header was already sent from somewhere else

Really? Is this in the current 2.0 branch, or in latest stable? That's a bug if that's happening (https://github.com/aidantwoods/SecureHeaders/wiki/csp#behaviour-when-a-csp-header-has-already-been-set)

**Comment by franzliedke** [comment]
Looks like that was just my bad memory then. ;)

**Comment by aidantwoods** [comment]
Okay, so that should add `SameSite` cookies and the new header that we're setting be default

**Comment by aidantwoods** [comment]
I'm a little conscious that `AUTO_ALL` probably wasn't the best name for the default in `auto` (in case that we want to add new settings that aren't on by default). So I'll probably change it to `AUTO_DEFAULT` before 2.0 goes out (something I realised while adding `SameSite`)

**Comment by aidantwoods** [comment]
Okay, so with those last commits adding back in the functionality missing from where this PR bases from, I've just copy pasted the current versions of the files to resolve the conflicts ;-) (mind doing a sanity check on that?) (don't worry about the regex that changed at the bottom, just want to make sure I didn't remove anything you've added by accident aha)

**Comment by aidantwoods** [comment]
Looks like PHP 5.4 isn't too happy with `array`s being set as constants, I guess we can change it to a `static` (I think that should be marginally less memory expensive than swapping it for just a regular variable?)

File: src/SecureHeaders.php [hunk 6]
```diff
@@ -875,6 +923,43 @@ private function pipeline()
...
+                $sameSite = 'Strict';
+            }
+            elseif ( ! isset($this->sameSiteCookies))
+            {
+                $sameSite = 'Lax';
+            }
+            else
+            {
+                $sameSite = $this->sameSiteCookies;
+            }
```
Comment (franzliedke) [inline src/SecureHeaders.php]: Can you extract both this if/elseif/else block and the second part of the large if condition above it into methods with meaningful names?

**Comment by franzliedke** [comment]
Looks very good! :)

Should we remove the `removeHeader()` method now as well? That feature doesn't really interact with the rest at all, or does it?

File: src/SecureHeaders.php (hunk 6 above)
Comment (aidantwoods) [reply src/SecureHeaders.php]: Yup, good call :)

**Comment by aidantwoods** [comment]
> Should we remove the removeHeader() method now as well? That feature doesn't really interact with the rest at all, or does it?

Isn't that still working, looks like it's part of the `pipeline` function to me?

**Comment by franzliedke** [comment]
Yes, but it's not really relevant to the responsibility of this class, which is to configure security headers. I am proposing to remove the `removeHeader()` method as well as the corresponding pipeline operation.

**Comment by aidantwoods** [comment]
Okay, yeah I see what you're saying. I guess the use-case is to implement a header blacklist (but this is nothing the equivalent adapter `header_remove` function couldn't do), so I suppose you're right: it's probably not needed anymore.

File: src/SecureHeaders.php (hunk 6 above)
Comment (aidantwoods) [reply src/SecureHeaders.php]: Looks like that large part of the `if` condition was me being paranoid anyway (I already enforced valid values in the setter).

**Comment by aidantwoods** [comment]
Okay, just reverted that push – looks like I forgot about the automatically added header removal use case 😉

**Comment by aidantwoods** [comment]
e.g.

```
$headers->removeHeader('X-XSS-Protection');
```

**Comment by franzliedke** [comment]
Ah true, good point.

**Comment by aidantwoods** [comment]
I can't really see any value in restricting its use to the automatically added headers either, so might as well allow it remove anything by keeping it as is

**Comment by aidantwoods** [comment]
I guess the notices safe mode produces are on the TODO list?

File: src/SecureHeaders.php [hunk 7]
```diff
@@ -1246,30 +1189,32 @@ private function deconstructHeaderValue(
         return $attributes;
     }
 
-    private function validateHeaders()
+    private function validateHeaders(HeaderBag $headers)
```
Comment (aidantwoods) [inline src/SecureHeaders.php:1188]: I don't think this works (CSP needs to be special cased in `Header` as the attribute decomposition algorithm differs)

Comment (franzliedke) [reply src/SecureHeaders.php:1188]: But it is? `$headers->forEachNamed('content-security-policy', ...)` - that is the special case.

Comment (aidantwoods) [reply src/SecureHeaders.php:1188]: The actual attribute decomposition isn't doing what you expect (i.e. the later calls to `$header->hasAttribute('report-uri')` etc.. won't work because even when `report-uri` exists in the CSP, it isn't being pulled out in the way that the code expects)

Comment (aidantwoods) [reply src/SecureHeaders.php:1188]: i.e `parseAttributes` needs to be special cased for CSP

Comment (aidantwoods) [reply src/SecureHeaders.php:1188]: I'll patch it up (can reuse some of the code in CompileCSP by making the policy decomposition static 😄 )

Comment (aidantwoods) [reply src/SecureHeaders.php:1188]: Okay, the validation process now produces notices correctly :)

**Comment by aidantwoods** [comment]
Haven't quite fully implemented the CSPHeader methods yet, just pushing to verify you think this is a sensible way to go? 😜

**Comment by aidantwoods** [comment]
Okay, I've added an `InjectStrictDynamic` operation, needed to be injected after the CSP is compiled (and merged), so that a true check of the outgoing CSP can be performed to decide on whether it is injectable.

**Comment by franzliedke** [comment]
Looks good from a quick glance.

The only thing I dislike is the decorator magic for the header classes. A `HeaderInterface` and a factory (method) that returns the correct instance for a given header type would be cleaner IMO.

**Comment by aidantwoods** [comment]
Yeah, it does always feel a bit dirty using magic methods for everything ;p

`Header` is now the interface name, and have added `HeaderFactory` for generating instances

**Comment by aidantwoods** [comment]
Right, so I think what's left to do is: warnings generated by operations, and more tests (though that'll probably be an ongoing goal).

**Comment by franzliedke** [comment]
Great! Do you still want to do that as part of this PR or can we get it merged?

Asking because I still want to do two things (in separate PRs):
- final cleanup (e.g. replace `done` with `apply($adapter)`)
- PSR-7 adapter

I'll leave the tests for you. You're the domain expert. Though you might want to write them for the master branch first, and then cherry-pick them onto 2.0 when you're done, so that we can see whether bugs are regressions.

**Comment by aidantwoods** [comment]
I'd probably like to get the warnings/notices done here (just so we're not losing anything on merge). Tests can wait though :) Writing them for master is probably a good idea 👍  

For any bugs that are non-regressions, I'll try patch them up and do a final release on `1.x` (just so there's a slightly better last version for PHP 5.3)

**Comment by aidantwoods** [comment]
RE the warnings/notices, ideally these would be output after the headers have been passed off to the `HttpAdapter` (and headers have actually been sent by PHP).

To that end, I wonder whether the Operation implementation should require a `collectErrors` method that SecureHeaders can use to collect any output generated by the `modify` method.

Once everything is stored in SecureHeaders, it's then a case of deciding how to best make sure errors are output *after* headers have been sent. Perhaps this role can be delegated to the `HttpAdapter`, since that is responsible for interacting with a framework (if any).

**Comment by franzliedke** [comment]
Two thoughts:

1. If at all possible, I would prefer to keep the warnings out of the operation classes, to keep these responsibilities separate. I remember you moved one of the warnings into one of them - was that a warning because of the existing headers'  contents, because of the faulty configuration (using the `SecureHeaders` facade), or a combination of the two? For the latter two cases, it should be easy to move the check back to `SecureHeaders`. Even for the third, it could be checked for based on the initial `HeaderBag` that's passed to the `apply` method.
2. How about returning a `Response` or `Result` object from the `apply` method - this could either be a successful response or one with warnings (the latter could even have methods to turn itself into a string or to pretty-print itself).

**Comment by aidantwoods** [comment]
1. I think it may end up being appropriate to have deciding which errors to output (as a result of the operation) be the responsibility of the `Operation`. In some cases (like with `ApplySafeMode`) it may be easy to extrapolate whether a warning is necessary based on whether certain header values changed, but what if later a single header being changed requires one of two notices?

   For the case of `InjectStrictDynamic`, it'll only output a warning if it could not inject `'strict-dynamic'` as a result of a missing nonce or hash in the CSP. But it will also leave the header unchanged if the directive already contains `'strict-dynamic'`, or contains the `'none'` keyword. (To differentiate between these cases externally, you'd have to re-implement all the checks the operation has already done to decide why it took a particular action, and whether a warning is necessary).

   I suppose we can make having an error collection method optional with a `method_exists` check though – to reduce clutter in operations that don't need to output anything.

2. I like this idea as a feature 👍 my only concern would be that it would mean that errors would essentially be off by default (devs now have to look for them). Having SecureHeaders be noisy by default means that you can teach devs something when they don't know they should configure something/they configure something badly.

**Comment by aidantwoods** [comment]
So I think we can probably put a pin in rethinking how errors are reported (just wanted to make sure notices/warnings weren't lost in this PR).

I'm happy to merge this now :) (unless you have any comments on my changes?)

**Comment by franzliedke** [comment]
Not much to add, other than I would slightly prefer an interface for `OperationWithErrors` rather than an abstract base class, but that's not too important...

**Comment by aidantwoods** [comment]
Hmm, we can finalise the specifics when we address errors later – though if it was an interface, it would require implementations to implement pretty much identical methods. Whereas with an abstract base, extensions can accept the default, or optionally specify a non default implementation – but an implementation is required (so no loss from an interface's requirements).

**Comment by franzliedke** [comment]
Thanks for merging! We're making real progress... 🎉 

I was more concerned about *depending* on an interface, rather than some abstract base class, that is just more flexible for the consumer of the interface in the future (because different implementations will be possible, with no changes to existing code necessary). That that abstract base class happens to implement two more trivial methods is not as important. As [duplication is far cheaper than the wrong abstraction](https://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction).

But yeah, it's not the biggest of issues. ;)

**Comment by aidantwoods** [comment]
> Thanks for merging! We're making real progress... 🎉

No worries! Thanks for all the great PRs! 😄 

> I was more concerned about depending on an interface, rather than some abstract base class...

We can properly address this when we revisit the errors 😉 if we name the interface `OperationWithErrors` and just have an abstract base class which implements this named something like `OperationWithNormalErrors` having `final` inheritable methods then we can probably address both our concerns :-)
